packages/micromark-util-events-to-acorn/dev/index.js

```diff
@@ -175,14 +227,38 @@ export function eventsToAcorn(events, options) {
       // @ts-expect-error: acorn has positions.
       esnode.range = [esnode.start, esnode.end]
     })
+
+    if (commentHandler) {
+      for (const comment of comments) {
+        commentHandler(comment)
+      }
```

Comment:
I think it’s easier to understand what’s going on by inlining the behavior:

```js
      if (Array.isArray(acornOptions.onComment)) {
        acornOptions.onComment.push(...comments)
      } else {
        for (const comment of comments) {
          acornOptions.onComment(
            type === 'Block',
            value,
            start,
            end,
            loc && loc.start,
            loc && loc.end
          )
        }
      }
```

So, instead of a `getCommentHandler` function in the otuer scope and calling it here, having the behavior where it’s needed.

I think inlining the behavior will also help deal with tokens:

```js
      if (Array.isArray(acornOptions.onToken)) {
        acornOptions.onToken.push(...tokens)
      } else {
        for (const token of tokens) {
          acornOptions.onToken(token)
        }
      }
```

With that, we can fix token positions *only* when a function/array is given, and you don’t need to add them to the tree:

```js
for (const token of tokens) {
  fixPositionOfToken(token) // To do: implement this
}

if (Array.isArray(acornOptions.onToken)) { /* ... */ } else { /* ... */ }
```

Reply:
`onComment/onToken` can be `undefined`, so I do have `if (commentHandler/tokenHandler)` check before iterating `comments/tokens`, and I'm trying to reuse the `visitor` logic for `tokens` at the same time to treat them just like `nodes`, and restore their types after.

And also, your proposed solution would increase the [`complexity`](https://eslint.org/docs/rules/complexity) which is complained by ESLint.